core/src/impl/Kokkos_Core.cpp

```diff
@@ -799,6 +806,74 @@ void parse_environment_variables(InitArguments& arguments) {
 
 }  // namespace
 
+int get_ctest_gpu(const char* local_rank_str)
+{
+  auto const* ctest_kokkos_device_type = std::getenv("CTEST_KOKKOS_DEVICE_TYPE");
+  if (!ctest_kokkos_device_type) {
+    return 0;
+  }
+
+  auto const* ctest_resource_group_count_str = std::getenv("CTEST_RESOURCE_GROUP_COUNT");
+  if (!ctest_resource_group_count_str) {
+    return 0;
+  }
+
+  // Make sure rank is within bounds of resource groups specified by CTest
+  auto resource_group_count = std::atoi(ctest_resource_group_count_str);
+  auto local_rank = std::atoi(local_rank_str);
+  if (local_rank >= resource_group_count) {
+    return 0;
+  }
+
+  // Get the resource types allocated to this resource group
+  std::ostringstream ctest_resource_group;
+  ctest_resource_group << "CTEST_RESOURCE_GROUP_" << local_rank;
+  auto const* ctest_resource_group_str = std::getenv(ctest_resource_group.str().c_str());
```

Comment:
How does that work?
`std::ostringstream::str()` returns by value.  Isn't the pointer returned by `c_str()` dangling?

Reply:
Good catch

Reply:
(I am not 100% sure)

Reply:
Actually it looks like the lifetime of the temporary is extended until the evaluation of the expression so it was fine :)

Reply:
No, you're right, the pointer returned by `c_str()` would have outlived the lifetime of the string returned by `str()`, which returns a copy rather than a reference. I have explicitly assigned it to a string now.